CORENET-5972: Add openvswitch-ipsec package into ipsec plugin

pperiyasamy opened the pull request:
This PR adds openvswitch3.5-ipsec package support to the ipsec os extension and ensures it integrates correctly for both node install and upgrade scenarios.

- Adds openvswitch3.5-ipsec to the ipsec os extension’s package list in the supported extension map.
- Adds proper ordering for openvswitch-ipsec.service: starts after ipsec.service, required before crio and kubelet
- Updates wait-for-ipsec-connect.service to start openvswitch-ipsec.service immediately when it is already configured for OVS, then waits for IKE SA establishment.
- Removes the previous workaround in ipsec-connect-wait that manually triggered pluto to establish IKE SAs.
- Introduces `LegacyExtensionPackages()` to track prior package lists for extensions whose packages changed from 4.x to 5.x. During upgrade, the new MCD pod may roll out before the OS image update, verification logic now accepts either the current or legacy package set, preventing nodes from degrading during the transition.
- Removes the early return in applyExtensions so that package-level changes within an extension are detected and applied via rpm-ostree.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

## Bug Fixes
- Improved Open vSwitch + IPsec startup ordering and readiness behavior, ensuring the IPsec-related service is enabled/started reliably when Open vSwitch is configured via `openshift.conf`.
- Enhanced IPsec connection readiness by deriving expected connections from Open vSwitch/OVN runtime state instead of static config parsing.
- Improved post-reboot extension package verification to accept both current and legacy valid IPsec package sets, including the new `openvswitch3.5-ipsec` option.

## Tests
- Added coverage for discovering valid package sets per extension and validating legacy IPsec upgrade scenarios.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

pperiyasamy committed:
Add service dependencies for openvswitch-ipsec.service

The openvswitch-ipsec service must be started only after ipsec.service
and it has to be up before crio and kubelet service, so add appropriate
systemd service dependencies for openvswitch-ipsec.service.

It also removes a workaround that was needed in ipsec-connect-wait
script to explicitly trigger pluto to establish IPSec IKE SAs with
peer nodes when openvswitch-ipsec.service is already running on
the node.

Signed-off-by: Periyasamy Palanisamy <pepalani@redhat.com>

pperiyasamy commented:
/retest-required

pperiyasamy commented:
/retest-required

pperiyasamy committed:
Add openvswitch3.5-ipsec package into ipsec plugin

The ipsec os extension is updated to install openvswitch3.5-ipsec
package in addition to libreswan and NetworkManager-libreswan
packages, so this commit updates supported extension map with
openvswitch-ipsec package.

Signed-off-by: Periyasamy Palanisamy <pepalani@redhat.com>

pperiyasamy commented:
/assign @igsilya

pperiyasamy commented:
/assign @jcaamano

pperiyasamy commented:
/retest

pperiyasamy commented:
/assign @huiran0826 @anuragthehatter @tssurya

igsilya commented on templates/common/_base/files/wait-for-ipsec-connect.yaml:
```diff
@@ -9,25 +9,39 @@ contents:
     establishedsa=""
     while [[ $elapsed -lt $timeout ]]; do
-      desiredconn=$(grep -E '^\s*conn\s+' /etc/ipsec.d/openshift.conf | grep -v '%default' | awk '{print $2}' | tr ' ' '\n' | sort | tr '\n' ' ')
+      if systemctl is-active --quiet $ovsipsecsvc; then
+        # When openvswitch-ipsec.service is active (i.e. IPsec is enabled for
+        # east west traffic), the service may be just started before
+        # wait-for-ipsec-connect.service, it may take a while to populate
+        # /etc/ipsec.d/openshift.conf with IPsec connection entries.
+        # so retrieve connection names from ovs.
+        desiredconn=$(ovs-vsctl --columns=name find interface type=geneve | awk -F': ' '/name/ {print $2}' | sort | awk '{print $0"-in-1", $0"-out-1"}' | tr '\n' ' ')
+      else
+        desiredconn=$(grep -E '^\s*conn\s+' /etc/ipsec.d/openshift.conf | grep -v '%default' | awk '{print $2}' | tr ' ' '\n' | sort | tr '\n' ' ')
+      fi
```
The ovs-vsctl method should work in both cases, so maybe just always use it instead of parsing openshift.conf?

igsilya replied:
If ipsec is disabled in OVN, then those tunnels should not have ipsec configuration.   Can we filter by `remote_cert` being non-empty, for example?  Something like:
```
ovs-vsctl --columns=name find interface type=geneve options:remote_cert!='""'
```
?

pperiyasamy replied:
yes @igsilya, filtering tunnels on `remote_name` would handle both cases, thanks ! updated the script.

pperiyasamy commented:
/retest

pperiyasamy committed:
Start openvswith-ipsec systemd service

When openvswith-ipsec systemd service is already configured for OVN, then let
wait-for-ipsec-connect.service start it immediately and then start waiting
for IKE SA to be established.

Signed-off-by: Periyasamy Palanisamy <pepalani@redhat.com>

jcaamano commented:
/cc @jcaamano

pperiyasamy commented:
/retest

pperiyasamy commented:
/test ?

pperiyasamy commented:
/reopen

pperiyasamy commented:
/remove-lifecycle rotten

pperiyasamy committed:
Support legacy extension packages during upgrade

Add openvswitch3.5-ipsec to the ipsec extension and implement graceful
handling of extension package changes during upgrade.

During an upgrade, the new MCD code may roll out before the OS image
update. If an extension’s package list has changed, this creates a
window where the new MCD expects packages that aren’t yet installed,
causing nodes to become degraded.

Introduce LegacyExtensionPackages() to track previous package lists
and modify verification logic to accept either current or legacy
package sets. This allows nodes to pass verification during the
upgrade transition period.

AssistedBy: Claude Sonnet 4.5 <noreply@anthropic.com>
Signed-off-by: Periyasamy Palanisamy <pepalani@redhat.com>

pperiyasamy commented on pkg/controller/common/helpers.go:
```diff
@@ -574,6 +574,39 @@ func SupportedExtensions() map[string][]string {
+// During an upgrade the new MCD code may run before the OS image update. This function
+// ensures that verification accepts either the old package set (from the current OS
+// image) OR the new package set (from the updated code).
+func GetAllValidPackageSetsForExtension(extension string) [][]string {
+	var allSets [][]string
+
+	// Add current package set
+	supported := SupportedExtensions()
+	if current, exists := supported[extension]; exists {
+		allSets = append(allSets, current)
+	}
+
+	// Add legacy package set (if exists)
+	legacy := LegacyExtensionPackages()
+	if legacy, exists := legacy[extension]; exists {
+		allSets = append(allSets, legacy)
+	}
+
+	return allSets
+}
```
addressed it with 4518a3a0d5da10e5de7ca8a120ba4b74757ebf3f.

djoshy commented on pkg/daemon/update.go:
```diff
@@ -1923,39 +1959,43 @@ func (dn *CoreOSDaemon) verifyExtensionPackages(config *mcfgv1.MachineConfig) er
-		return fmt.Errorf("failed to query RPM database for package %q: %v: %s", pkg, err, strings.TrimSpace(string(out)))
-	}
+		// Try to match against any valid package set
+		var matchedSet []string
```
This seems to be only used as a bool? Could we change it to one?

pperiyasamy replied:
sure, done.

djoshy commented on pkg/controller/common/helpers.go:
```diff
@@ -574,6 +574,39 @@ func SupportedExtensions() map[string][]string {
+	// Add legacy package set (if exists)
+	legacy := LegacyExtensionPackages()
+	if legacy, exists := legacy[extension]; exists {
+		allSets = append(allSets, legacy)
```
Not a blocking issue, but the naming is a little confusing here

pperiyasamy replied:
updated the variable names now, is that better now ?

djoshy commented on pkg/controller/common/helpers_test.go:
```diff
@@ -2002,3 +2002,136 @@ func TestDetectRuncInMachineConfig(t *testing.T) {
+	}
+}
+
+func TestLegacyExtensionPackageUpgradeScenario(t *testing.T) {
```
This test only validates the data in `GetAllValidPackageSetsForExtension`(already covered by your other test), it doesn't seem to be actually calling the daemon's upgrade flow? Not sure that's easy to do in a unit either way, I would at least want `areAllPackagesInstalled` called to mock the upgrade path.

pperiyasamy replied:
moved `TestLegacyExtensionPackageUpgradeScenario` test into `pkg/daemon/update_test.go` to validate `areAllPackagesInstalled`.

djoshy reviewed (commented):
This still doesn't cover the pre-reboot scenario - we should also update `verifyExtensionsStaged` to use the new helpers.

Some of the extensions units seem to be broken, which will need to be fixed.

pperiyasamy commented:
> This still doesn't cover the pre-reboot scenario - we should also update `verifyExtensionsStaged` to use the new helpers.
>  

@djoshy Had this change for `verifyExtensionsStaged` with commit https://github.com/openshift/machine-config-operator/commit/d17142e0335bc7e6d54f36d7d4cbdf8b282e1bb4, but had to restore it back to address the comment https://github.com/openshift/machine-config-operator/pull/4878#discussion_r3477524795. isn't that a valid scenario for a fresh installation ? we shouldn't end up in staging old package list.

> Some of the extensions units seem to be broken, which will need to be fixed.

Thanks, fixed it.

FYI. just tested 4.22 to 5.0 ipsec upgrade lane (see below job link) with this PR and it is installing openvswitch3.5-ipsec package now.
https://prow.ci.openshift.org/view/gs/test-platform-results/logs/multi-pr-openshift-cluster-network-operator-2662-openshift-ovn-kubernetes-3254-openshift-machine-config-operator-4878-e2e-aws-ovn-upgrade-ipsec/2071515885629083648

djoshy commented:
> > This still doesn't cover the pre-reboot scenario - we should also update verifyExtensionsStaged to use the new helpers.
>
> @djoshy Had this change for verifyExtensionsStaged with commit https://github.com/openshift/machine-config-operator/commit/d17142e0335bc7e6d54f36d7d4cbdf8b282e1bb4, but had to restore it back to address the comment https://github.com/openshift/machine-config-operator/pull/4878#discussion_r3477524795. isn't that a valid scenario for a fresh installation ? we shouldn't end up in staging old package list.

`verifyExtensionsStaged` runs pre-reboot during install or upgrade, so if we want to only have guard it during the upgrade, we will have to pass down the `firstBoot` boolean through to the `verifyExtensionsStaged` helper. I believe it is currently passed in the update daemon call: https://github.com/openshift/machine-config-operator/blob/f05e48737d9db0a62f45f6ef06fc7382a4333fed/pkg/daemon/update.go#L894

It will need some threading through for it be passed to `applyExtensions`...which might not look pretty, but I'm OK with it if you feel strongly about making the check upgrade only.

Thinking about this a bit more, does `verifyExtensionPackages` even need a legacy check? If this function is only called post node reboot, we won't run into the new deamon x old packages list problem in that path. Perhaps we can just have the legacy check in `verifyExtensionsStaged`, then?

We should also make a follow-up card to remove the legacy extension check in 5.1 altogether, since all clusters would only be upgrading from 5.0.x, and so all nodes should have the new packages already.

pperiyasamy commented:
> > > This still doesn't cover the pre-reboot scenario - we should also update verifyExtensionsStaged to use the new helpers.
> > 
> > 
> > @djoshy Had this change for verifyExtensionsStaged with commit [d17142e](https://github.com/openshift/machine-config-operator/commit/d17142e0335bc7e6d54f36d7d4cbdf8b282e1bb4), but had to restore it back to address the comment [#4878 (comment)](https://github.com/openshift/machine-config-operator/pull/4878#discussion_r3477524795). isn't that a valid scenario for a fresh installation ? we shouldn't end up in staging old package list.
> 
> `verifyExtensionsStaged` runs pre-reboot during install or upgrade, so if we want to only have guard it during the upgrade, we will have to pass down the `firstBoot` boolean through to the `verifyExtensionsStaged` helper. I believe it is currently passed in the update daemon call:
> 
> https://github.com/openshift/machine-config-operator/blob/f05e48737d9db0a62f45f6ef06fc7382a4333fed/pkg/daemon/update.go#L894
> 
> It will need some threading through for it be passed to `applyExtensions`...which might not look pretty, but I'm OK with it if you feel strongly about making the check upgrade only.
> 

yes, I think the current change in `applyExtensions` runs rpm-ostree command with the diff packages for the upgrade (openvswitch3.5-ipsec package in this scenario), the `verifyExtensionsStaged` is invoked bit later and i guess doesn't require to be validated against old package set. You could check logs from Loki to confirm this.

For example:

```
2026-06-29 09:57:54	
{"container":"machine-config-daemon","host":"ip-10-0-16-77.us-west-1.compute.internal","pod":"machine-config-daemon-rhrb6","_entry":"I0629 09:57:54.228745    2599 update.go:3167] \"Applying extensions : [\\\"update\\\" \\\"--install\\\" \\\"NetworkManager-libreswan\\\" \\\"--install\\\" \\\"libreswan\\\"]\""}
2026-06-29 10:00:06	
{"container":"machine-config-daemon","host":"ip-10-0-16-77.us-west-1.compute.internal","pod":"machine-config-daemon-rhrb6","_entry":"I0629 09:57:54.228745    2599 update.go:3167] \"Applying extensions : [\\\"update\\\" \\\"--install\\\" \\\"NetworkManager-libreswan\\\" \\\"--install\\\" \\\"libreswan\\\"]\""}
2026-06-29 10:00:11	
{"boot_id":"850125c347244cbd915bcf6a2d2894e1","host":"ip-10-0-16-77","systemd_unit":"crio-c260c95d08edf371a4231cf4196b78d310440d2929d86c4684374b5ac918fc0a.scope","_entry":"machine-config-daemon[2599]: \"Applying extensions : [\\\"update\\\" \\\"--install\\\" \\\"NetworkManager-libreswan\\\" \\\"--install\\\" \\\"libreswan\\\"]\""}
2026-06-29 11:36:01	
{"container":"machine-config-daemon","host":"ip-10-0-16-77.us-west-1.compute.internal","pod":"machine-config-daemon-wqcrr","_entry":"I0629 11:36:00.972192   80737 update.go:3175] \"Applying extensions : [\\\"update\\\" \\\"--install\\\" \\\"openvswitch3.5-ipsec\\\"]\""}
2026-06-29 11:36:01	
{"boot_id":"8c1d710fee7947e9ab60a7375d51c9ff","host":"ip-10-0-16-77","systemd_unit":"crio-3d0f483f09efae458d93cf78033c9becec4453c44167978c44a6b136def0fe04.scope","_entry":"machine-config-daemon[80737]: \"Applying extensions : [\\\"update\\\" \\\"--install\\\" \\\"openvswitch3.5-ipsec\\\"]\""}
2026-06-29 11:38:47	
{"boot_id":"850125c347244cbd915bcf6a2d2894e1","host":"ip-10-0-16-77","systemd_unit":"crio-c260c95d08edf371a4231cf4196b78d310440d2929d86c4684374b5ac918fc0a.scope","_entry":"machine-config-daemon[2599]: \"Applying extensions : [\\\"update\\\" \\\"--install\\\" \\\"NetworkManager-libreswan\\\" \\\"--install\\\" \\\"libreswan\\\"]\""}
2026-06-29 11:38:51	
{"container":"machine-config-daemon","host":"ip-10-0-16-77.us-west-1.compute.internal","pod":"machine-config-daemon-wqcrr","_entry":"I0629 11:36:00.972192   80737 update.go:3175] \"Applying extensions : [\\\"update\\\" \\\"--install\\\" \\\"openvswitch3.5-ipsec\\\"]\""}
2026-06-29 11:38:54	
{"boot_id":"8c1d710fee7947e9ab60a7375d51c9ff","host":"ip-10-0-16-77","systemd_unit":"crio-3d0f483f09efae458d93cf78033c9becec4453c44167978c44a6b136def0fe04.scope","_entry":"machine-config-daemon[80737]: \"Applying extensions : [\\\"update\\\" \\\"--install\\\" \\\"openvswitch3.5-ipsec\\\"]\""}
```

> Thinking about this a bit more, does `verifyExtensionPackages` even need a legacy check? If this function is only called post node reboot, we won't run into the new deamon x old packages list problem in that path. Perhaps we can just have the legacy check in `verifyExtensionsStaged`, then?
> 

This legacy packages check seems to be still necessary on `verifyExtensionPackages` ,  Tried with a test [PR](https://github.com/openshift/machine-config-operator/pull/6250), machine-config upgrade fails and mc pools go into degrade state (because openvswitch3.5-ipsec package not found from RPM database). Please take a look this [job](https://prow.ci.openshift.org/view/gs/test-platform-results/logs/multi-pr-openshift-cluster-network-operator-2662-openshift-ovn-kubernetes-3254-openshift-machine-config-operator-6250-e2e-aws-ovn-upgrade-ipsec/2071974446561759232) for the logs.

> We should also make a follow-up card to remove the legacy extension check in 5.1 altogether, since all clusters would only be upgrading from 5.0.x, and so all nodes should have the new packages already.

Shouldn't this change required for 5.x versions as well ? because user might upgrade from 4.x to any 5.x version. is that correct ?

djoshy commented:
ack, I see thanks for explaining - then we should be able to just have the legacy check in `verifyExtensionPackages`. 

> Shouldn't this change required for 5.x versions as well ? because user might upgrade from 4.x to any 5.x version. is that correct ?

Currently, you're aren't allowed to skip level upgrade as far as I know. So you'd have to go through 5.0 to get to 5.1 and so on. Thinking about this a bit though, there are customers who do EUS upgrades by pausing the worker pool, so it is possible you'd have worker nodes jumping a few OCP versions. Perhaps we leave this as is.

pperiyasamy commented:
/payload 5.0 ci blocking
/payload 5.0 nightly blocking

pperiyasamy commented:
/payload-job periodic-ci-openshift-hypershift-release-5.0-periodics-e2e-aks
/payload-job periodic-ci-openshift-hypershift-release-5.0-periodics-e2e-aws-ovn
/payload-job periodic-ci-openshift-release-main-ci-5.0-e2e-aws-ovn-techpreview
/payload-job periodic-ci-openshift-release-main-nightly-5.0-e2e-metal-ipi-ovn-ipv4

pperiyasamy commented:
/pipeline required

jcaamano commented on templates/common/_base/files/wait-for-ipsec-connect.yaml:
```diff
@@ -4,33 +4,68 @@ contents:
     establishedsa=""
     while [[ $elapsed -lt $timeout ]]; do
-      desiredconn=$(grep -E '^\s*conn\s+' /etc/ipsec.d/openshift.conf | grep -v '%default' | awk '{print $2}' | tr ' ' '\n' | sort | tr '\n' ' ')
+      if ! desiredconn=$(ovs-vsctl --columns=name find interface type=geneve options:remote_name!='""' \
```
Use `ovs-vsctl` option `--bare`

pperiyasamy replied:
done.

jcaamano commented on templates/common/_base/files/wait-for-ipsec-connect.yaml:
```diff
@@ -4,33 +4,68 @@ contents:
+    # When node is not already running with openvswitch-ipsec systemd
+    # service, then explicitly load the OVN IPsec connections to
+    # establish IKE SAs.
+    if ! systemctl is-active --quiet $ovsipsecsvc; then
```
I don't understand that we start the service and wait for it to have started, and then immediately after we check if the service is inactive. Shouldn't both blocks be coordinated by some kind of if/else pattern?
```
if grep -q "openshift.conf" /etc/sysconfig/openvswitch; then
  ...
else
  ...
```
I guess we are not expecting the service to become inactive between both blocks because that would be racey.

pperiyasamy replied:
you are right, it can be handled with if-else patten. the if block is for openvswitch-ipsec is already configured for ipsec, so enable and start it. the else block is for transitioning period where openvswitch-ipsec service is not configured yet, so tweak openshift.conf file to load ipsec connections explicitly.

jcaamano replied:
We agreed to check the return code of `systemctl enable --now $ovsipsecsvc` only, no need to check if the service is active later on.

jcaamano commented on templates/common/_base/files/wait-for-ipsec-connect.yaml:
```diff
@@ -4,33 +4,68 @@ contents:
+        exit 1
+      fi
       establishedsa=$(ipsec showstates | grep ESTABLISHED_CHILD_SA | grep -o '"[^"]*"' | sed 's/"//g' | tr ' ' '\n' | sort | uniq | tr '\n' ' ')
       if [ "$desiredconn" == "$establishedsa" ]; then
         echo "IPsec SAs are established for desired connections after ${elapsed}s"
```
Should we `exit 1` if `"$desiredconn" != "$establishedsa"`?

pperiyasamy replied:
We don't need `exit 1` because script just wanted to establish all IKE SAs if possible just before kubelet service comes up. but don't want to return exit code 1 if that fails.

jcaamano replied:
We agreed to exit 1 on any unexpected outcome for visbility.

pperiyasamy replied:
done

igsilya replied:
FWIW, this condition will not be reached when one of the nodes in the cluster is down for some reason.  I.e. if you shut down one node, this service will exit with 1 on reboot of any other nodes.

A bulk or a cascade reboot of multiple nodes also will have a high chance for this service to exit 1 on most of the rebooted nodes, because reboots can take many minutes and different amount of time on different nodes.  Not sure if this is bad or not, but just to highlight the potential confusing behavior for the users who will see the service failure.

jcaamano replied:
The alternative is to hide other potential issues.

If we think it is common and reasonable to expect this, let's change it back but let's add a comment on which scenarios this is expected.

igsilya replied:
I'm not sure what's the best course of actions here.  We may leave it as is and revisit if it actually causes any confusion among users.  Should be an easy fix.

pperiyasamy replied:
sure, we mainly introduced this service to avoid disruptions in CI for the ipsec upgrade job. so not sure how much this service is going to be looked upon by a customer. of course ,we will come back and fix it if needed. thanks.

pperiyasamy replied:
had a discussion with @jcaamano, we decided  to do `exit 0` now with a comment. updated with the fix.

jcaamano commented on templates/common/_base/files/wait-for-ipsec-connect.yaml:
```diff
@@ -4,33 +4,68 @@ contents:
+        | awk '{print $0"-in-1", $0"-out-1"}' \
+        | tr '\n' ' '); then
+        echo "failed to query OVS Geneve interfaces for desired IPsec connections"
+        exit 1
```
So when something unexpected happens, we `exit 1`. However when this service fails it does not prevent the node from continue its boot process with a non-functioning ipsec setup, right? I guess this is a conscious design choice but we should have a comment of its reasoning somewhere here.

pperiyasamy replied:
we don't have to fail the service if this command fails (i don't think this ever happen), but changed with `exit 0` now.

jcaamano replied:
We agreed to add docs to explain this is a best effort to bootstrap ipsec from the get go avoiding the disruptions we could potentially have otherwise if delayed.

pperiyasamy replied:
done, added relevant info for this.

jcaamano commented on pkg/daemon/update.go:
```diff
@@ -1817,9 +1817,8 @@ func (dn *CoreOSDaemon) applyExtensions(oldConfig, newConfig *mcfgv1.MachineConf
-	if (extensionsEmpty) ||
-		(reflect.DeepEqual(oldConfig.Spec.Extensions, newConfig.Spec.Extensions) && oldConfig.Spec.OSImageURL == newConfig.Spec.OSImageURL) {
+	// If both configs have no extensions, there's nothing to do
+	if len(oldConfig.Spec.Extensions) == 0 && len(newConfig.Spec.Extensions) == 0 {
```
Why do you actually need this change? Wouldn't an update to `BaseOSExtensionsContainerImage` allowing us to install new packages always happen at the same time of an update to `OSImageURL`?
Doesn't this change open a small window where if a reconciliation happens between daemon update but before reboot, it would fail when trying to install a new package because we would still be running with the old `BaseOSExtensionsContainerImage`?

pperiyasamy replied:
this change is to detect package updates within extensions (though no changes with `.Spec.Extensions` itself). after the update of new mcp pod, it applies the new package just before [node](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/multi-pr-openshift-cluster-network-operator-2662-openshift-ovn-kubernetes-3254-openshift-machine-config-operator-4878-e2e-aws-ovn-upgrade-ipsec/2074088584977059840/artifacts/e2e-aws-ovn-upgrade-ipsec/gather-extra/artifacts/nodes/ip-10-0-114-145.us-west-1.compute.internal/) reboot. have not seen any issue due to mismatch with `BaseOSExtensionsContainerImage` (or) `OSImageURL`.

`Jul 06 13:51:51.972577 ip-10-0-114-145 root[74161]: machine-config-daemon[61580]: "Applying extensions : [\"update\" \"--install\" \"openvswitch3.5-ipsec\"]" `

do you still want to check on `OSImageURL` ?

jcaamano replied:
Talked with @peri offline and we don't have a good understanding on what kind of change we need here.

On one side we would expect OSImageURL to change on a upgrade and all extension packages to be layered again. But on testing upgrade [jobs](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/multi-pr-openshift-cluster-network-operator-2662-openshift-ovn-kubernetes-3254-openshift-machine-config-operator-4878-e2e-aws-ovn-upgrade-ipsec/2074088584977059840/) we are not seeing that happening hence this kind of change.

On the other side this kind of change would be required anyway if we just want to add new extension packages through MCO upgrade without upgrading OS.

I guess that as long as the new packages are available in the old BaseOSExtensionsContainerImage this should not be an issue other than potentially incurring in noop install attempts? 
If new packages are not available in old BaseOSExtensionsContainerImage then this might introduce a small window were an unlikely machine config change between MCO update and OS update can trigger a failed install?

Anyway, a lot of lucubration on my part, hope that @igsilya @djoshy can chime in.

djoshy replied:
I'm not sure changing this condition is actually required for this [particular run](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/multi-pr-openshift-cluster-network-operator-2662-openshift-ovn-kubernetes-3254-openshift-machine-config-operator-4878-e2e-aws-ovn-upgrade-ipsec/2074088584977059840/) to succeed, as the OSImageURL actually changed in this case. It does help in upgrade scenarios where the `OSImageURL`/`Extensions` does not change(but the list of packages within an already installed extension does change). Maybe this happened in earlier CI runs, prompting @pperiyasamy to add it? From a customer POV though, this is quite unlikely to take place. 

That said though, we should also change how [mcDiff.extensions](https://github.com/openshift/machine-config-operator/blob/86e75381a6fe51bc8dfac969fa2a4d4fc0528971/pkg/daemon/update.go#L1566) is computed to match this. This is inturn used in `applyOSChanges`'s outer gate (`mcDiff.osUpdate || mcDiff.extensions || mcDiff.kernelType || mcDiff.kargs || mcDiff.oclEnabled`). If we don't, the current fix will only cover a narrow set of circumstances. 

> I guess that is good as long as the new packages are available in the old BaseOSExtensionsContainerImage this should not be an issue other than potentially incurring in noop installs attempts? If new packages are not available in old BaseOSExtensionsContainerImage then this might introduce a small windows were an unlikely machine config change betwee MCO update and OS update can trigger a failed install?

`OSImageURL` and `BaseOSExtensionsContainerImage` are always [fetched together](https://github.com/openshift/machine-config-operator/blob/86e75381a6fe51bc8dfac969fa2a4d4fc0528971/pkg/operator/sync.go#L1987) from the same `ConfigMap`, and manifest ordering guarantees the new MCD always comes after that `ConfigMap` is updated, so a new MCD can never be paired with the old images.

cc @yuqi-zhang in case I missed something here!

yuqi-zhang replied:
I agree with what David is saying. This change is essentially a no-op right now, and even if you were to have relaxed the outer check, in practice I think all it would accomplish is that if the user ever does a MachineConfig change themselves without changing os contents, it would unnecessarily remove and reapply extensions, which then I think also results in a no-op on rpm-ostree side.

For the scenario you're trying to guard against to occur, the user would have to:
1. be already using the extension
2. undergo an upgrade, where there is 0 changes to the osImageURL

Which essentially can never happen unless you plan to backport this and it hits a z-stream where somehow nothing gets rebuilt, and the user happens to do a .z-1 to .z upgrade for those specific versions. I would lean towards not changing this logic.

pperiyasamy replied:
thanks all for your inputs, tested 4.22 to 5.0 upgrade after reverting the commit and ipsec extension is updated correctly with `openvswitch3.5-ipsec` package, now i really don't remember what triggered me to have that change (may be it's to make upgrade work from 5.0 stable build -> PR version), this is not valid scenario for customer. all good now. removed the commit.

pperiyasamy commented:
the below unit test failed, not related to PR changes, running it again.

```
=== NAME  TestAssertions/Runs_and_is_canceled_with_a_context
        	Error Trace:	/go/src/github.com/openshift/machine-config-operator/test/helpers/assertions_test.go:175
```

pperiyasamy commented:
/test unit

pperiyasamy commented:
@igsilya @djoshy need your +1 review on this. PTAL.

jcaamano commented on templates/common/_base/files/wait-for-ipsec-connect.yaml:
```diff
@@ -4,33 +4,62 @@ contents:
+        fi
+      done
+    else
+      echo "$ovsipsecsvc is not running, so make pluto to establish IKE SAs"
```
tweak this echo a bit to say its not configured instead of running

pperiyasamy replied:
done.

pperiyasamy commented:
/pipeline required

igsilya commented:
@pperiyasamy I'm not an expert in the extension packages management parts, but the systemd service related code seems correct to me, if we're OK with the oneshot waiting service reported as failed in relatively common cases (multi-node reboots).

jcaamano commented:
lgtm

pperiyasamy commented:
/retest

jcaamano commented:
/lgtm

igsilya commented:
I have no power in this repo, but /lgtm as well.

pperiyasamy commented:
/retest

djoshy commented:
/lgtm

pperiyasamy commented:
/verified in ci

pperiyasamy commented:
/verified by ci

jcaamano commented:
/retest-required
